gather-scheduler.js

```diff
@@ -18,99 +18,114 @@
-
-      // Gather: setup phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.setup(options)))
-
-      // Enable tracing and network record collection.
-      .then(_ => driver.beginTrace())
-      .then(_ => driver.beginNetworkCollect())
-
-      // Gather: beforePageLoad phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.beforePageLoad(options)))
-
-      // Load the page (if the CLI / extension want it loaded).
-      .then(_ => {
-        if (loadPage) {
-          return driver.gotoURL(url, driver.WAIT_FOR_LOADED);
-        }
-
-        return Promise.resolve();
-      })
+      return driver.forceUpdateServiceWorkers();
+    });
+  }
 
-      // Gather: afterPageLoad phase
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.afterPageLoad(options)))
+  // Enable tracing and network record collection.
+  static beginPassiveCollection(driver) {
+    return driver.beginTrace().then(_ => {
+      return driver.beginNetworkCollect();
+    });
+  }
 
-      // Disable network collection; grab records.
-      .then(_ => driver.endNetworkCollect())
-      .then(networkRecords => {
-        tracingData.networkRecords = networkRecords;
-      })
+  static endPassiveCollection(driver, gatherers, options, tracingData) {
+    return driver.endNetworkCollect().then(networkRecords => {
+      tracingData.networkRecords = networkRecords;
+    }).then(_ => {
+      return driver.endTrace();
+    }).then(traceContents => {
+      tracingData.traceContents = traceContents;
+    });
+  }
 
-      // Disable tracing; grab records.
-      .then(_ => driver.endTrace())
-      .then(traceContents => {
-        tracingData.traceContents = traceContents;
-      })
+  static _phase(phaseName) {
+    return function(driver, gatherers, options, tracingData) {
+      return GatherScheduler._runPhase(gatherers, gatherer => {
+        return gatherer[phaseName](options, tracingData);
+      });
+    };
+  }
 
-      // Gather: afterTraceCollected phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.afterTraceCollected(options, tracingData)))
+  static _runPhase(gatherers, gatherFun) {
+    return gatherers.reduce((chain, gatherer) => {
+      return chain.then(_ => gatherFun(gatherer));
+    }, Promise.resolve());
+  }
 
-      // Disconnect the driver.
-      .then(_ => driver.disconnect())
+  static run(gatherers, options) {
+    const driver = options.driver;
+    const tracingData = {};
+    const self = GatherScheduler;
 
-      // Gather: tearDown phase.
-      .then(_ => this._runPhase(gatherers,
-        gatherer => gatherer.tearDown(options, tracingData)))
+    if (options.url === undefined || options.url === null) {
+      throw new Error('You must provide a url to scheduler');
+    }
 
+    return GatherScheduler.runSeries([
+      driver.connect.bind(driver),
+      self.setupDriver,
+      self._phase('setup'),
```

Comment:
we can iterate on this more, I felt that this was an OK first step.. you are correct none of these are side effect free or re-entrant (however I do not think these can ever be since we do not have state free exits). The goal here was the make the list of steps readable, and I feel pretty ok with what this ended up being like.

I felt that adding the new steps onto the .then chain pushed it over the limit of readability.


I want to iterate more on this and we can continue to tweak